crates/ide/src/inlay_hints/param_name/generic.rs

```diff
@@ -0,0 +1,288 @@
+//! Implementation of inlay hints for generic parameters.
+use ide_db::{active_parameter::generic_def_for_node, RootDatabase};
+use syntax::{
+    ast::{self, HasName},
+    AstNode, SyntaxNode,
+};
+
+use crate::{inlay_hints::GenericParameterHints, InlayHint, InlayHintsConfig, InlayKind};
+
+use super::{is_argument_similar_to_param_name, render_label};
+
+pub(crate) fn hints(
+    acc: &mut Vec<InlayHint>,
+    sema: &hir::Semantics<'_, RootDatabase>,
+    config: &InlayHintsConfig,
+    node: SyntaxNode,
+) -> Option<()> {
+    if matches!(config.generic_parameter_hints, GenericParameterHints::Never) {
+        return None;
+    }
+
+    let const_generics_only =
+        matches!(config.generic_parameter_hints, GenericParameterHints::ConstOnly);
+
+    if let Some(generic_arg_list) = ast::GenericArgList::cast(node.clone()) {
+        let (generic_def, _, _, _) =
+            generic_def_for_node(sema, &generic_arg_list, &node.first_token()?)?;
+
+        let mut args = generic_arg_list.generic_args().peekable();
+        let start_with_lifetime = matches!(args.peek()?, ast::GenericArg::LifetimeArg(_));
+        let params = generic_def.params(sema.db).into_iter().filter(|p| {
+            if let hir::GenericParam::TypeParam(it) = p {
+                if it.is_implicit(sema.db) {
+                    return false;
+                }
+            }
+            if !start_with_lifetime {
+                return !matches!(p, hir::GenericParam::LifetimeParam(_));
+            }
+            true
+        });
+
+        let hints = params.zip(args).filter_map(|(param, arg)| {
+            if matches!(arg, ast::GenericArg::AssocTypeArg(_)) {
+                return None;
+            }
+            let source_syntax = match param {
+                hir::GenericParam::TypeParam(it) => {
+                    if const_generics_only {
+                        return None;
+                    }
+                    sema.source(it.merge())?.value.syntax().clone()
+                }
+                hir::GenericParam::ConstParam(it) => {
+                    let syntax = sema.source(it.merge())?.value.syntax().clone();
+                    let const_param = ast::ConstParam::cast(syntax)?;
+                    const_param.name()?.syntax().clone()
+                }
+                hir::GenericParam::LifetimeParam(it) => {
+                    if const_generics_only {
+                        return None;
+                    }
+                    sema.source(it)?.value.syntax().clone()
+                }
+            };
```

Comment:
Querying the source can be expensive, so lets move this down to where we actually need it so we give the other filtering things a run first